Feature/rsa signatures

shawnjk opened the pull request:
Implements a feature to sign OAuth requests using RSA-SHA1.

Commit:
Implement RSA-SHA1 signatures for iOS

Commit:
Add tests for RSA signatures

Commit:
Add CFRetain and CFRelease statements around BDBOAuth1RequestSerializer initializers

Commit:
Switch RSAPrivateKey property to id from SecKeyRef for ARC

smholloway commented on BDBOAuth1Manager/BDBOAuth1RequestSerializer.m:
```diff
@@ -321,6 +339,45 @@ - (NSString *)OAuthSignatureForMethod:(NSString *)method
+    uint8_t signature[signatureSize];
+    SecKeyRawSign((__bridge SecKeyRef)self.RSAPrivateKey, kSecPaddingPKCS1SHA1, digest, sizeof(digest), signature, &signatureSize);
+#if (defined(__IPHONE_OS_VERSION_MIN_REQUIRED) && __IPHONE_OS_VERSION_MIN_REQUIRED >= 70000) || (defined(__MAC_OS_X_VERSION_MIN_REQUIRED) && __MAC_OS_X_VERSION_MIN_REQUIRED >= 1090)
+
```
Extra blank line here?

smholloway commented on BDBOAuth1Manager/BDBOAuth1RequestSerializer.m:
```diff
@@ -214,6 +215,25 @@ - (instancetype)initWithService:(NSString *)service
++ (instancetype)serializerForService:(NSString *)service
+                     withConsumerKey:(NSString *)consumerKey
+                       RSAPrivateKey:(id)RSAPrivateKey {
+    return[[[self class] alloc] initWithService:service consumerKey:consumerKey RSAPrivateKey:RSAPrivateKey];
```
Space after `return` please.

smholloway commented on BDBOAuth1Manager/BDBOAuth1RequestSerializer.m:
```diff
@@ -287,7 +307,12 @@ - (NSDictionary *)OAuthParameters {
     parameters[BDBOAuth1SignatureTimestampParameter]   = [@(floor([[NSDate date] timeIntervalSince1970])) stringValue];
-    parameters[BDBOAuth1SignatureMethodParameter]      = @"HMAC-SHA1";
+
+    if (self.RSAPrivateKey) {
```
I forget the truth table in Objective-C--this will be true if the `RSAPrivateKey` is initialized and not `0`?

shawnjk replied:
Yes, unfortunately I cannot find a good reference Objective C truthy table though.

smholloway commented on BDBOAuth1ManagerDemo/BDBOAuth1ManagerTests-OSX/BDBOAuth1ManagerTests_OSX.m:
```diff
@@ -0,0 +1,64 @@
+    XCTAssertEqualObjects(signature, expectedSignature, @"Computed OAuth signature does not match expected signature");
+}
+
+
```
Extra blank line.

smholloway commented on BDBOAuth1ManagerDemo/BDBOAuth1ManagerTests-OSX/BDBOAuth1ManagerTests_OSX.m:
```diff
@@ -0,0 +1,64 @@
+}
+
+
+@end
```
Adding tests == pro. Well done.

smholloway commented on BDBOAuth1ManagerDemo/Podfile:
```diff
@@ -2,8 +2,15 @@ source 'https://github.com/CocoaPods/Specs.git'
+link_with 'BDBOAuth1ManagerTests-OSX'
+
+pod 'BDBOAuth1Manager', :path => '../'
+end
```
I don't know Pods--can you explain this addition (lines 10-16), please?

shawnjk replied:
Basically, the Pods are different based on the platform, iOS or OSX, so this creates another set of Pods linked to my OSX test target.

smholloway replied:
Why would you test against OS X instead of iOS?

shawnjk replied:
I have tests for both. I just wanted to regression test the OSX HMAC signing because RSA signing only works for iOS.

smholloway replied:
Ahhh, okay. Right on.

Commit:
Remove extra empty lines

Commit:
Remove more empty lines

Commit:
Fix missing OAuth parameters with RSA signatures

Commit:
Upgrade AFNetworking to 2.6.0

Commit:
URL encode / and ? in the query string

AFNetworking 2.6 removed those characters from its encoded character
list. See https://github.com/AFNetworking/AFNetworking/pull/2908

Commit:
Merge pull request #2 from toopher/feature/upgrade-afnetworking

Feature/upgrade afnetworking

bdbergeron commented:
Hi guys! First of all, thanks for using this project and making it even better!

I was just curious as to whether or not you'd like to merge this feature back up into the main project? It definitely seems like something that other users might want to have in the library. If you'd like to submit a PR, I'd be happy to merge it in!

shawnjk commented:
@bdbergeron we'd love to merge back into main. Created a PR: https://github.com/bdbergeron/BDBOAuth1Manager/pull/39
